### PR TITLE
speedup installer

### PR DESCRIPTION
**Pull request by Erotemic**
The exact details of this PR will need discussion, but the core concepts I think are good. This introduces a few optimizations to speedup install time. Improved sections are:

* faster install calls to the apt package manager when prereqs are already installed.
* faster install of python packages with uv.

The first optimization is only for debian-based systems, but it saves a lot of time if you go through repeat installs. It uses dpkg to check if you already have the packages installed, and then just skips the call to `apt install` if everything is already setup. It uses my `apt_ensure` bash function which is part of my [dotfile utilities](https://github.com/Erotemic/local/blob/main/init/utils.sh). Details of the implementation and suggestions for where it should live in or be refactored for this code base should be discussed. For now, I put it in common.sh

The second optimization is to use uv instead of pip to install packages. This requires using pip to install uv, but its worth the setup cost. Subsequent installs go much faster and make up for the time. The use of uv can be configured with a variable, but it is hard-coded for now.

There is also another feature hard-coded to false, but when debugging the installer I didn't want to keep re-creating my ovos-installer venv, so I added an option to disable starting fresh, but it defaults to false and there is no way to enable it as I'm not sure exactly where config options like this should be handled, but I thought it was worth bringing up. 

Lastly, this PR contains questions on a few things in the code. Namely, why install to /root/.ansible instead of `$RUN_AS_HOME/.ansible`. If possible I would like to minimize any use of sudo in the installer, and ideally make it optional as long as you've setup your system beforehand. 

## Summary by CodeRabbit

- **New Features**
	- Introduced new environment variables for enhanced installation options: `USE_UV` for faster package management and `REUSED_CACHED_ARTIFACTS` for retaining existing virtual environments.
	- Added a new function to check and install missing packages without requiring elevated permissions.

- **Improvements**
	- Streamlined the installation process for required packages and virtual environments, improving overall efficiency and flexibility.

### CODE REVIEW

**Comment by goldyfruit** [comment]
Hi @Erotemic

Could you please check the CodeRabbit's recommendations?
Most of them makes sense in IMHO.

Thanks :+1:

File: utils/common.sh
Comment (Erotemic) [reply utils/common.sh]: I am not sure if this is correct or not, I would need to test it. I forget the exact semantics of what bash is doing here, but shellcheck often complains about similar things when the way I wrote it is intended. Not sure if that is the case here or not. Will look into it later.

File: utils/common.sh
Comment (Erotemic) [reply utils/common.sh]: I need maintainer input on this. The AI is just suggesting what I was asking about. I do think it is better to use this, but I'm wondering if maintainers have a reason for writing as `/root/.ansible`.

File: utils/common.sh
Comment (Erotemic) [reply utils/common.sh]: I'm not sure about this. What is the difference here?

File: utils/common.sh
Comment (Erotemic) [reply utils/common.sh]: I think the original is ok. What is the problem with it? Can you provide an example where it fails?

File: utils/common.sh
Comment (Erotemic) [reply utils/common.sh]: Please provide an example where the old version fails, but the new version works correctly.

File: utils/common.sh
Comment (Erotemic) [reply utils/common.sh]: Upon further consideration, I like the AI version better. It makes sense to use the more robust version of array length that can handle arguments with spaces in case someone uses this code as an example for something where spaces can exist. Can you redo the suggestion so I can apply it?

**Comment by goldyfruit** [comment]
> @goldyfruit I do need maintainer input on the last unresolved one in regarding `/root/.ansible` vs `"$RUN_AS_HOME/.ansible"`. I think the latter is probably better, but I want to make sure there isn't a special reason why the former is used.

That could be a separated PR as its a bit out of scope of this one (performances improvement).

**Comment by goldyfruit** [comment]
> * Does `apt_ensure` live in common.sh? Or would it make sense to make a new place for it to live?

Since we only have one for now, I would say it can stay here until we do the same for `pacman`, `dnf` and `zypper`.

> * The `apt_ensure` documentation style is different than other functions. It uses a **doc** variable to make it docstring-like. Do maintainers like this style, or should I just use comments above the function?

I would keep the same documentation style for now. We could have a different PR handling all the functions 
 documentation.

> * The new config variables are just plopped into `create_python_venv`, is there a better way to handle upfront configuration? Perhaps these go in argparse.sh? It seems like these probably should just be environment variable and not necessarily added to the CLI args. Does it make sense to just put variables like that into argparse.sh and have it control everything configuration wise.

I would handle that in `argparse.sh` with two new arguments.

**Comment by goldyfruit** [comment]
image
The `pip` messages should be redirected to the log file.

File: utils/common.sh
Comment (Erotemic) [reply utils/common.sh]: I believe `uv` is available in any environment that ovos supports, and the default of true is where much of the speedup in this PR comes from. 

Defaulting to false would be safer, but I think we should default it to true and then see if anyone raises an issue. My guess is that it won't cause any, but if it does it is easy to flip it back to false. I'll defer to maintainers on what the default should be, but I think there is a good argument to go with "true".

**Comment by Erotemic** [comment]
I fixed the docs formatting. There isn't a need to go to a `__doc__` style, I just like it because it matches my Python inclinations.

I've moved the definition of the new environment variables to argparse, but I didn't make them CLI flags, I kept them as pure environment variables. My reasoning is that these are not normal things a user would want to change as part of the normal install UX, but they are useful to expose to an advanced user comfortable with environment variables that might be debugging the system.

I fixed the logging issue. That was an oversight.

In a discussion with coderabbitai, it suggested defaulting USE_UV to false, but I think I made a decent argument for why it should default to true, but I will yield to maintainer opinion here.

**Comment by goldyfruit** [comment]
> In a discussion with coderabbitai, it suggested defaulting USE_UV to false, but I think I made a decent argument for why it should default to true, but I will yield to maintainer opinion here.

I agree, lets turn it to `true`, worst case scenario we turn it back to `false`. :+1:

**Review by goldyfruit — APPROVED** [review]

**Comment by goldyfruit** [comment]
> I fixed the docs formatting. There isn't a need to go to a `__doc__` style, I just like it because it matches my Python inclinations.

I think you didn't push the commit.

**Comment by Erotemic** [comment]
That's very strange. I don't see the commit here, but my local checkout is telling me that the remote is up to date. 
image

I also see the commits on [my fork](https://github.com/OpenVoiceOS/ovos-installer/compare/main...Erotemic:ovos-installer:speedup-installer). Going to try rebasing on main to see if that helps. (And it did seem to work).

**Review by goldyfruit — APPROVED** [review]

**Comment by goldyfruit** [comment]
Thanks @Erotemic, great PR :+1:

**Pull request merged**
